Project root node of @if and @for blocks

crisbeto opened the pull request:
A couple of fixes to get the content projection behavior of `@if` and `@for` blocks closer to `*ngIf` and `*ngFor`. Includes the following commits:

### fix(compiler): project control flow root elements into correct slot
With the directive-based control flow users were able to conditionally project content using the `*` syntax. E.g. `<div *ngIf="expr" projectMe></div>` will be projected into `<ng-content select="[projectMe]"/>`, because the attributes and tag name from the `div` are copied to the template via the template creation instruction. With `@if` and `@for` that is not the case, because the conditional is placed *around* elements, rather than *on* them. The result is that content projection won't work in the same way if a user converts from `*ngIf` to `@if`.

These changes aim to cover the most common case by doing the same copying when a control flow node has *one and only one* root element or template node.

This approach comes with some caveats:
1. As soon as any other node is added to the root, the copying behavior won't work anymore. A diagnostic will be added to flag cases like this and to explain how to work around it.
2. If `preserveWhitespaces` is enabled, it's very likely that indentation will break this workaround, because it'll include an additional text node as the first child. We can work around it here, but in a discussion it was decided not to, because the user explicitly opted into preserving the whitespace and we would have to drop it from the generated code. The diagnostic mentioned point 1 will flag such cases to users.

### refactor(compiler): implement control flow content projection fix in template pipeline
Recreates the fix for content projection in control flow in the new template pipeline. I also had to make the following adjustments to the pipeline:
1. The `TemplateOp.tag` property was being used to generate the name of the template function, rather than the actual tag name being passed into `ɵɵtemplate`. Since the content projection fix requires the tag name to be passed in, I've introduced a new `functionNameSuffix` property instead.
2. `TemplateOp.block` was being used to determine whether to pass `TemplateOp.tag` into the `ɵɵtemplate` instruction. Now that we're always passing in the tag name after the refactor in point 1, we no longer need this flag.

In addition to the refactors above, I also made some minor cleanups where I saw the opportunity to do so.

Fixes #52277.

crisbeto committed:
fix(compiler): project control flow root elements into correct slot

With the directive-based control flow users were able to conditionally project content using the `*` syntax. E.g. `<div *ngIf="expr" projectMe></div>` will be projected into `<ng-content select="[projectMe]"/>`, because the attributes and tag name from the `div` are copied to the template via the template creation instruction. With `@if` and `@for` that is not the case, because the conditional is placed *around* elements, rather than *on* them. The result is that content projection won't work in the same way if a user converts from `*ngIf` to `@if`.

These changes aim to cover the most common case by doing the same copying when a control flow node has *one and only one* root element or template node.

This approach comes with some caveats:
1. As soon as any other node is added to the root, the copying behavior won't work anymore. A diagnostic will be added to flag cases like this and to explain how to work around it.
2. If `preserveWhitespaces` is enabled, it's very likely that indentation will break this workaround, because it'll include an additional text node as the first child. We can work around it here, but in a discussion it was decided not to, because the user explicitly opted into preserving the whitespace and we would have to drop it from the generated code. The diagnostic mentioned point #1 will flag such cases to users.

Fixes #52277.

crisbeto commented on packages/compiler/src/template/pipeline/ir/src/ops/create.ts:
```diff
@@ -179,11 +179,9 @@ export interface TemplateOp extends ElementOpBase {
+   * Suffix to add to the name of the generated template function.
    */
-  block: boolean;
+  functionNameSuffix: string|null;
```
I'm not super attached to the name so I'm open to suggestions. The main goal was to stop using the tag name for the function name generation.

crisbeto commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -149,10 +148,6 @@ function ingestElement(unit: ViewCompilationUnit, element: t.Element): void {
     throw Error(`Unhandled i18n metadata type for element: ${element.i18n.constructor.name}`);
   }
 
-  const staticAttributes: Record<string, string> = {};
```
Unrelated change to the fix, I just noticed that this map wasn't being used anywhere.

dylhunn replied:
Makes sense, we refactored this into attribute extraction phases a while back.

crisbeto commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -14,7 +14,6 @@ import {splitNsName} from '../../../ml_parser/tags';
 import * as o from '../../../output/output_ast';
 import {ParseSourceSpan} from '../../../parse_util';
 import * as t from '../../../render3/r3_ast';
-import {Identifiers} from '../../../render3/r3_identifiers';
```
Unrelated change, the import wasn't used.

dylhunn replied:
I think I don't have import cleaning on, thanks.

crisbeto commented on packages/compiler/src/template/pipeline/src/instruction.ts:
```diff
@@ -235,22 +235,23 @@ export function i18nStart(slot: number, constIndex: number, subTemplateIndex: nu
-      args.push(o.variable(emptyViewFnName));
-      args.push(o.literal(emptyDecls));
-      args.push(o.literal(emptyVars));
+      args.push(o.variable(emptyViewFnName), o.literal(emptyDecls), o.literal(emptyVars));
```
Unrelated change, I saw that we can avoid some function calls by pushing all the arguments at once.

crisbeto commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -2,7 +2,7 @@
     "uncompressed": {
       "runtime": 4252,
-      "main": 507632,
+      "main": 512758,
```
Not quite sure where this diff came from, but we were likely on the edge of the payload limit.

JeanMeche replied:
fwiw #52234 thas the same

crisbeto commented:
I've pushed some changes to avoid future issues if we start capturing comments by default and to rebase.

pkozlowski-opensource reviewed (commented):
LGTM overall but I would like to revisit this assumption:

> If `preserveWhitespaces` is enabled, it's very likely that indentation will break this
 *    workaround, because it'll include an additional text node as the first child. We can work
 *    around it here, but in a discussion it was decided not to, because the user explicitly opted
 *    into preserving the whitespace and we would have to drop it from the generated code.
 *    The diagnostic mentioned point #1 will flag such cases to users.
 
 I would say that we should have just one rule: we opt out of the root node capturing if there is more than one root node. This will be more common case with `preserveWhitespaces` but still there is one consistent rule here.

pkozlowski-opensource commented on packages/core/test/acceptance/control_flow_for_spec.ts:
```diff
@@ -277,4 +278,269 @@ describe('control flow - for', () => {
+      expect(fixture.nativeElement.textContent).toBe('Main: Before  After Slot: 123');
+    });
+
+    it('should not project the root node when preserveWhitespaces is enabled', () => {
```
"should not project the root node when preserveWhitespaces is enabled" => "should not project the root node when preserveWhitespaces is enabled and there are whitespace nodes" or something along those lines - currently it could be interpreted as opting out of the tag capturing every time `preserveWhitespaces` is used.

crisbeto commented:
Feedback has been addressed.

pkozlowski-opensource reviewed (approved):
LGTM, thnx!

dylhunn commented on packages/compiler/src/template/pipeline/src/phases/naming.ts:
```diff
@@ -77,15 +77,14 @@ function addNamesToView(
               state, compatibility);
         }
-        const repeaterToken =
-            op.tag === null ? '' : '_' + prefixWithNamespace(op.tag, op.namespace);
```
I kind of preferred having this prefixing logic in the naming phase as opposed to ingest, where you've now folded it into the new suffix field -- it seems like it's logically part of naming. It's a weak preference though. I don't think we could conceivably need to use the raw suffix for anything else, so it's probably fine?

crisbeto replied:
I agree that it would've felt more appropriate to have the naming logic there, but the rules we have for the function naming are a bit too different for the template nodes vs control flow that I couldn't find a better place than this one. I think that once we remove the definition builder, we can revisit the naming scheme.

dylhunn reviewed (approved):
LGTM. Makes perfect sense to me, and thank you for also updating the template pipeline. A lot of the use of `tag` for control flow and conditionals was marginal to begin with -- it was more designed for elements and containers.

crisbeto committed:
refactor(compiler): implement control flow content projection fix in template pipeline

Recreates the fix for content projection in control flow in the new template pipeline. I also had to make the following adjustments to the pipeline:
1. The `TemplateOp.tag` property was being used to generate the name of the template function, rather than the actual tag name being passed into `ɵɵtemplate`. Since the content projection fix requires the tag name to be passed in, I've introduced a new `functionNameSuffix` property instead.
2. `TemplateOp.block` was being used to determine whether to pass `TemplateOp.tag` into the `ɵɵtemplate` instruction. Now that we're always passing in the tag name after the refactor in point 1, we no longer need this flag.

In addition to the refactors above, I also made some minor cleanups where I saw the opportunity to do so.

crisbeto commented:
Caretaker note: this was passing presubmit until an hour ago. I've just rebased it which reset the status.

alxhub commented:
This PR was merged into the repository by commit c7b730eb721eec7f6a2f34dead2fff8088f3d15b.
